pkg/engine/compat.go

```diff
@@ -42,110 +42,186 @@ type streamsResultBuilder struct {
 	streams map[string]int
 	data    logqlmodel.Streams
 	count   int
+
+	// buffer of rows to be reused between calls to CollectRecord to reduce reallocations of slices and builders
+	rowsBuffer rowsBuffer
 }
 
-func (b *streamsResultBuilder) CollectRecord(rec arrow.Record) {
-	for row := range int(rec.NumRows()) {
-		stream, entry := b.collectRow(rec, row)
+type rowsBuffer struct {
+	len              int
+	timestamps       []time.Time
+	lines            []string
+	lbsBuilders      []*labels.Builder
+	metadataBuilders []*labels.Builder
+	parsedBuilders   []*labels.Builder
+}
 
-		// Ignore rows that don't have stream labels, log line, or timestamp
-		if stream.IsEmpty() || entry.Line == "" || entry.Timestamp.Equal(time.Time{}) {
-			continue
-		}
+func (p *rowsBuffer) prepareFor(newLen int) {
+	if newLen <= p.len {
+		p.timestamps = p.timestamps[:newLen]
+		clear(p.timestamps)
 
-		// Add the entry to the result builder
-		key := stream.String()
-		idx, ok := b.streams[key]
-		if !ok {
-			idx = len(b.data)
-			b.streams[key] = idx
-			b.data = append(b.data, push.Stream{Labels: key})
+		p.lines = p.lines[:newLen]
+		clear(p.lines)
+
+		p.lbsBuilders = p.lbsBuilders[:newLen]
+		p.metadataBuilders = p.metadataBuilders[:newLen]
+		p.parsedBuilders = p.parsedBuilders[:newLen]
```

Comment:
You may be leaking the builders from the previous row in the range `[newLen:]`.

Reply:
Fixed 🙇